File: packages/core/src/memory/bm25.ts
export type BM25Doc = { id: string; text: string };
export type BM25Result = { id: string; score: number };
type TokenizedDoc = { id: string; tokens: string[]; length: number };

function tokenize(text: string): string[] { return text.toLowerCase().split(/\W+/).filter(Boolean); }

const K1 = 1.5;
const B = 0.75;

export function createBM25Index(docs: BM25Doc[]) {
  const tokenizedDocs: TokenizedDoc[] = docs.map((d) => { const tokens = tokenize(d.text); return { id: d.id, tokens, length: tokens.length }; });
  const avgDocLength = tokenizedDocs.length > 0 ? tokenizedDocs.reduce((sum, d) => sum + d.length, 0) / tokenizedDocs.length : 0;
  const df = new Map<string, number>();
  for (const doc of tokenizedDocs) { const seen = new Set(doc.tokens); for (const token of seen) { df.set(token, (df.get(token) ?? 0) + 1); } }
  const N = tokenizedDocs.length;

  return {
    search(query: string): BM25Result[] {
      const queryTokens = tokenize(query);
      if (queryTokens.length === 0 || N === 0) return [];
      const results: BM25Result[] = [];
      for (const doc of tokenizedDocs) {
        let score = 0;
        const tf = new Map<string, number>();
        for (const token of doc.tokens) { tf.set(token, (tf.get(token) ?? 0) + 1); }
        for (const term of queryTokens) {
          const termFreq = tf.get(term) ?? 0;
          if (termFreq === 0) continue;
          const docFreq = df.get(term) ?? 0;
          const idf = Math.log((N - docFreq + 0.5) / (docFreq + 0.5) + 1);
          score += idf * ((termFreq * (K1 + 1)) / (termFreq + K1 * (1 - B + B * (doc.length / avgDocLength))));
        }
        if (score > 0) results.push({ id: doc.id, score });
      }
      return results.sort((a, b) => b.score - a.score);
    },
  };
}

File: v3/apps/pi_core/lib/pi_core/memory/bm25.ex
defmodule PiCore.Memory.BM25 do
  @moduledoc "BM25 keyword scoring for memory search."

  @k1 1.5
  @b 0.75

  def search(docs, query) when is_list(docs) do
    query_tokens = tokenize(query)
    if query_tokens == [] || docs == [], do: [], else: do_search(docs, query_tokens)
  end

  defp do_search(docs, query_tokens) do
    tokenized = Enum.map(docs, fn {id, text} ->
      tokens = tokenize(text)
      {id, tokens, length(tokens)}
    end)

    avg_len = if tokenized == [], do: 0,
      else: Enum.reduce(tokenized, 0, fn {_, _, len}, acc -> acc + len end) / length(tokenized)

    n = length(tokenized)

    # Document frequency
    df = Enum.reduce(tokenized, %{}, fn {_, tokens, _}, acc ->
      tokens |> MapSet.new() |> Enum.reduce(acc, fn token, a ->
        Map.update(a, token, 1, & &1 + 1)
      end)
    end)

    tokenized
    |> Enum.map(fn {id, tokens, doc_len} ->
      tf = Enum.frequencies(tokens)
      score = Enum.reduce(query_tokens, 0.0, fn term, acc ->
        term_freq = Map.get(tf, term, 0)
        if term_freq == 0, do: acc, else: acc + bm25_score(term_freq, doc_len, avg_len, Map.get(df, term, 0), n)
      end)
      {id, score}
    end)
    |> Enum.filter(fn {_, score} -> score > 0 end)
    |> Enum.sort_by(fn {_, score} -> -score end)
  end

  defp bm25_score(tf, doc_len, avg_len, doc_freq, n) do
    idf = :math.log((n - doc_freq + 0.5) / (doc_freq + 0.5) + 1)
    numerator = tf * (@k1 + 1)
    denominator = tf + @k1 * (1 - @b + @b * (doc_len / max(avg_len, 1)))
    idf * (numerator / denominator)
  end

  defp tokenize(text) do
    text
    |> String.downcase()
    |> String.split(~r/\W+/)
    |> Enum.filter(& &1 != "")
  end
end
